### ISSUE TITLE
ResNet50/152: DataRead Failure in batch normalization statistics.

### ISSUE THREAD

**Issue by bcbrock** [Microsoft/CNTK#1598]
Hi,

I'm trying to work through an example using the ResNet50 network on an
abbreviated ImageNet input set:

Examples/Image/Classification/ResNet/BrainScript/ResNet50_IMageNet1K.cntk

I am doing parallel training on a single system with 4 GPUs. I reduced the
training set to only 20 images per category to save time. I also need to
reduce the mini-batch size to avoid OOM errors on my GPUs. Training works
fine. However I am observing the following error in the `BNStatistics` phase:

```
[CALL STACK]
[0x106872bc]        Microsoft::MSR::CNTK::PostComputingActions<float>::  BatchNormalizationStatistics  (Microsoft::MSR::CNTK::IDataReader*,  std::vector<std::basic_string<wchar_t,std::char_traits<wchar_t>,std::allocator<wchar_t>>,std::allocator<std::basic_string<wchar_t,std::char_traits<wchar_t>,std::allocator<wchar_t>>>> const&,  std::basic_string<wchar_t,std::char_traits<wchar_t>,std::allocator<wchar_t>>,  unsigned long,  int) + 0x151c
[0x10520434]        void  DoBatchNormalizationStat  <float>(Microsoft::MSR::CNTK::ConfigParameters const&) + 0x844
[0x104b5160]        void  DoCommands  <float>(Microsoft::MSR::CNTK::ConfigParameters const&,  std::shared_ptr<Microsoft::MSR::CNTK::MPIWrapper> const&) + 0xe50
[0x10427a40]        wmainOldCNTKConfig  (int,  wchar_t**)              + 0xa30
[0x10428354]        wmain1  (int,  wchar_t**)                          + 0x134
[0x103ba7e0]        main                                               + 0x100
[0x3fffae22309c]                                                       + 0x2309c
[0x3fffae223298]    __libc_start_main                                  + 0xb8
EXCEPTION occurred: DataRead Failure in batch normalization statistics
```

This error occurs when the `minibatchSize` parameter is set to 100 both in
the `TrainNetwork` clause and the `BNStatistics` clause. However if
`minibatchSize` is set to 128 in both places no error is observed. I have also
observed errors and correct behavior in other cases but can't detect a
pattern.

After briefly looking into the code to see how this error is generated, my
suspicion is that there may be undocumented assumptions about the
relationships between the number of samples, minibatch sizes for training and
the batch normalization, and the `itersPerNode` parameter. If so, I would be
interested to learn what this relationship is, since as of now I can't find
any settings that allow me to run the ResNet152 network on my system without
the above (or other) errors.

**Comment by cha-zhang** [comment]
You can simply skip the BNStatistics step for now. This was a work we did a while ago and thought it is useful, but so far we don't have a firm proof yet.

**Comment by bcbrock** [comment]
Thank you.

**Comment by DemjanGrubic** [comment]
@cha-zhang I have seen quite some improvement with BNStatistics step on ResNet50 (3% on top-1 and 1.5% on top-5).
I am generating validation error vs epoch graphs with BNStatistics, and I am generating it as I described in the comment here: https://github.com/Microsoft/CNTK/issues/1035#issuecomment-286455773

I didn't have any problem with ResNet50, but on ResNet152 I am getting following error:
```
[CALL STACK]
[0x849c8c]                                                            
[0x7f5a6cbd066d]    Microsoft::MSR::CNTK::Matrix<float>::  VectorMax  (Microsoft::MSR::CNTK::Matrix<float>&,  Microsoft::MSR::CNTK::Matrix<float>&,  bool,  int) const + 0x19d
[0x98049a]          Microsoft::MSR::CNTK::ClassificationErrorNode<float>::  ForwardPropNonLooping  () + 0xfa
[0x8533a4]          Microsoft::MSR::CNTK::ComputationNodeNonLooping<float>::  ForwardProp  (Microsoft::MSR::CNTK::FrameRange const&) + 0x14
[0xbe84f7]          Microsoft::MSR::CNTK::ComputationNetwork::PARTraversalFlowControlNode::  ForwardProp  (Microsoft::MSR::CNTK::FrameRange const&) + 0x177
[0xbec36c]          Microsoft::MSR::CNTK::ComputationNetwork::  ForwardProp  (std::shared_ptr<Microsoft::MSR::CNTK::ComputationNodeBase>) + 0x7c
[0x9111f9]          void Microsoft::MSR::CNTK::ComputationNetwork::  ForwardProp  <std::vector<std::shared_ptr<Microsoft::MSR::CNTK::ComputationNodeBase>,std::allocator<std::shared_ptr<Microsoft::MSR::CNTK::ComputationNodeBase>>>>(std::vector<std::shared_ptr<Microsoft::MSR::CNTK::ComputationNodeBase>,std::allocator<std::shared_ptr<Microsoft::MSR::CNTK::ComputationNodeBase>>> const&) + 0x169
[0x927141]          Microsoft::MSR::CNTK::SimpleEvaluator<float>::  Evaluate  (Microsoft::MSR::CNTK::IDataReader*,  std::vector<std::basic_string<wchar_t,std::char_traits<wchar_t>,std::allocator<wchar_t>>,std::allocator<std::basic_string<wchar_t,std::char_traits<wchar_t>,std::allocator<wchar_t>>>> const&,  unsigned long,  unsigned long) + 0x12a1
[0x928faf]          void  DoEval  <float>(Microsoft::MSR::CNTK::ConfigParameters const&) + 0x7af
[0x89b6e5]          void  DoCommands  <float>(Microsoft::MSR::CNTK::ConfigParameters const&,  std::shared_ptr<Microsoft::MSR::CNTK::MPIWrapper> const&) + 0xe35
[0x81f955]          wmainOldCNTKConfig  (int,  wchar_t**)              + 0x9c5
[0x82007f]          wmain1  (int,  wchar_t**)                          + 0x7f
[0x7d1b68]          main                                               + 0xd8
[0x7f5a6afe9f45]    __libc_start_main                                  + 0xf5
[0x7dddde]                                                            
EXCEPTION occurred: VectorMax: Matrix is empty.
```

How can I solve this?

**Comment by cha-zhang** [comment]
What data set are you trained on? ImageNet?

**Comment by DemjanGrubic** [comment]
Yes, ImageNet 2012 (which should the same as 2013, 2014).

**Comment by cha-zhang** [comment]
OK, we will look into it.

**Comment by yuxiaoguo** [comment]
@bcbrock I will look into the potential problem you reported. 

@DemjanGrubic Do you use the recipe of ResNet50 and ResNet152 in the Examples/Image/Classification/ResNet/BrainScript?

**Comment by DemjanGrubic** [comment]
@yuxiaogou Yes, I am using those cntk configs.

**Comment by yuxiaoguo** [comment]
@DemjanGrubic In PBN stage, every evaluated BN node will export a line, like "Estimating Statistics --> XX node". Could you paste these lines that I can infer reason to cause your failure? From your log, the failure was happened in ClassificationError Node. The node should be never called in PBN, since we limit the maximum iterations. There may be something wrong to make the ClassificationError called.

**Comment by DemjanGrubic** [comment]
@yuxiaoguo I am sharing all log files from BNStatistics (these are last logs when I tried, minibatch size is 128 for BNStatistics, even though I trained with 256, but it is the same error as when minibatch size is equal to 256 (I tried this because of the first comment)):
[8bit_ResNet152_val.rank1.txt](https://github.com/Microsoft/CNTK/files/873558/8bit_ResNet152_val.rank1.txt)
[8bit_ResNet152_val.rank2.txt](https://github.com/Microsoft/CNTK/files/873562/8bit_ResNet152_val.rank2.txt)
[8bit_ResNet152_val.rank3.txt](https://github.com/Microsoft/CNTK/files/873559/8bit_ResNet152_val.rank3.txt)
[8bit_ResNet152_val.rank4.txt](https://github.com/Microsoft/CNTK/files/873561/8bit_ResNet152_val.rank4.txt)
[8bit_ResNet152_val.rank5.txt](https://github.com/Microsoft/CNTK/files/873563/8bit_ResNet152_val.rank5.txt)
[8bit_ResNet152_val.rank6.txt](https://github.com/Microsoft/CNTK/files/873560/8bit_ResNet152_val.rank6.txt)
[8bit_ResNet152_val.rank7.txt](https://github.com/Microsoft/CNTK/files/873564/8bit_ResNet152_val.rank7.txt)
[8bit_ResNet152_val.txt](https://github.com/Microsoft/CNTK/files/873565/8bit_ResNet152_val.txt)

Thanks,

**Comment by yuxiaoguo** [comment]
@DemjanGrubic Thanks for your feedback. It seems the PBN process has been finished and exception happens in the evaluation phase. Have you even tried to evaluate the model without PBN? Is the same exception triggered?

**Comment by DemjanGrubic** [comment]
@yuxiaoguo Yes, sorry for misleading about the step in which error occurrences.
I haven't tried evaluation without PBN because I had cvReader setup. But since evaluation gives the same error without PBN, that means the problem is with top5Errs evaluation node (it is evaluated only during Eval command). Once I removed that node from Eval, it works.

I don't know why that top5Errs node make a problem for Eval?

**Comment by yuxiaoguo** [comment]
@DemjanGrubic Never mind. I will loop someone to solve your problem. However, you may try Top1 error first in your case.

**Comment by DemjanGrubic** [comment]
@yuxiaoguo Yes, I am currently doing just top1. And also if someone could answer my comment from a question I linked (https://github.com/Microsoft/CNTK/issues/1035#issuecomment-286455773). Currently, I need to do PBN and Eval for every epoch's model which is really slow.
I am not even sure if there is a faster way?
Thanks for your help!
